### PR TITLE
[WOR-1193] Workspace deletion async runner

### PR DESCRIPTION
**Pull request by aherbst-broad**
[Ticket](https://broadworkbench.atlassian.net/browse/WOR-1193)

## Context
We are moving to an async model for workspace deletion. In this [PR](https://github.com/broadinstitute/rawls/commits/develop), I implemented the wireup of the API and creation of the monitor record that hands off work to a backend runner task. This PR implements the runner task to orchestrate the async workspace deletion.

## This PR
* The meat of the PR is in the WorkspaceDeletionRunner class and associated helper classes. This class starts the deletion jobs with downstream systems (Leo + WSM) and the eventual Rawls record deletion.
  * This is somewhat stateless approach. For each step we adjust the relevant job record in the DB to reflect the current step (i.e., WorkspaceDeletionInit, upon completion, will change the record job type to LeoAppDeletion). 
  * I broke out the Leo + WSM deletion code into their own classes for improved testability and hopefully clearer error handling.
* (Finally?) adds a `WorkspaceRepository` class, again to aid testability. It's contained to basic get/delete + updating state.
* I've put this implementation on a [BEE here](https://terra.aherbst-swatomation-sturgeon.bee.envs-terra.bio). Happy to add folks to my sandbox billing project for testing purposes. 

## Test Plan
* Pushed this change to a BEE
* Created a workspace with a runtime + apps (cromwell + wds)
* Issued a deletion request to the deletion V2 API
* Observed:
    * Apps deleted
    * Runtime deleted
    * WSM record deleted
    * Rawls record deleted 

---

**PR checklist**

- [ ] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should [publish a new official `rawls-model`](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model) and update `rawls-model` in [Orchestration's dependencies](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/project/Dependencies.scala).
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceManagerResourceMonitorRecordDao.scala [hunk 1]
```diff
@@ -13,10 +13,6 @@ case class WorkspaceManagerResourceMonitorRecordDao(dataSource: SlickDataSource)
   def selectAll(): Future[Seq[WorkspaceManagerResourceMonitorRecord]] =
     dataSource.inTransaction(_.WorkspaceManagerResourceMonitorRecordQuery.getRecords)
 
-  def selectByBillingProject(name: RawlsBillingProjectName): Future[Seq[WorkspaceManagerResourceMonitorRecord]] =
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/WorkspaceManagerResourceMonitorRecordDao.scala:16]: Unrelated; the IDE called this out as dead code.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/workspace/runners/deletion/actions/WsmDeletionAction.scala [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import org.broadinstitute.dsde.rawls.model.{RawlsRequestContext, Workspace}
+import org.broadinstitute.dsde.rawls.monitor.workspace.runners.deletion.actions.DeletionAction.when500
+import org.broadinstitute.dsde.rawls.util.Retry
+import org.broadinstitute.dsde.rawls.workspace.WorkspaceManagerOperationFailureException
+
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future, blocking}
+import scala.util.{Failure, Success, Try}
+
+class WsmDeletionAction(workspaceManagerDao: WorkspaceManagerDAO,
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/workspace/runners/deletion/actions/WsmDeletionAction.scala]: Much of this code can be made common w/the existing stuff in the MultiCloudWorkspaceService. I'm limiting the scope here but will ticket out that remediation.

File: project/Dependencies.scala [hunk 3]
```diff
@@ -132,7 +132,7 @@ object Dependencies {
   val billingProfileManager = excludeJakarta("bio.terra" % "billing-profile-manager-client" % "0.1.185-SNAPSHOT")
   val terraCommonLib = tclExclusions(excludeJakarta("bio.terra" % "terra-common-lib" % "0.0.93-SNAPSHOT" classifier "plain"))
   val sam: ModuleID = excludeJakarta("org.broadinstitute.dsde.workbench" %% "sam-client" % "0.1-d606036")
-  val leonardo: ModuleID = "org.broadinstitute.dsde.workbench" % "leonardo-client_2.13" % "1.3.6-8740f99-SNAP"
+  val leonardo: ModuleID = "org.broadinstitute.dsde.workbench" % "leonardo-client_2.13" % "1.3.6-c331e18-SNAP"
```
Comment (aherbst-broad) [inline project/Dependencies.scala:135]: There was a bug in leo's swagger and the associated client that was causing gson deserialization errors when calling `listAzureRuntimes`. This is a fixed version.

**Comment by aherbst-broad** [comment]
Closing for now; I'm reworking the approach as this will block downstream WSM monitor tasks for an unacceptably long time if a deletion takes O(minutes) (which it usually does).

**Comment by aherbst-broad** [comment]
jenkins retest

**Review by blakery — APPROVED** [review]

**Review by marctalbott — APPROVED** [review]

**Comment by aherbst-broad** [comment]
jenkins retest

**Pull request merged**